**Linux** [Build 7689](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=66602&buildTypeId=WFPR) outcome was **FAILURE** using a merge of 86282c09de21309a3c8df4ed8a1406ce5d7ec0cf
Summary: Tests failed: 2, passed: 3087, ignored: 364 Build time: 00:49:12
### Failed tests
```
org.jboss.as.test.integration.domain.mixed.eap630.MixedDomain630TestSuite: org.jboss.as.test.integration.domain.mixed.eap620.MixedDomain620TestSuite: java.lang.RuntimeException: java.lang.RuntimeException: MgmtOperationException{operation={
    "operation" => "remove",
    "address" => [
        ("profile" => "full-ha"),
        ("subsystem" => "jgroups"),
        ("stack" => "udp"),
        ("protocol" => "RSVP")
    ]
}, result={
    "outcome" => "failed",
    "failure-description" => {"domain-failure-description" => "WFLYCTL0216: Management resource '[
    (\"profile\" => \"full-ha\"),
    (\"subsystem\" => \"jgroups\"),
    (\"stack\" => \"udp\"),
    (\"protocol\" => \"RSVP\")
]' not found"},
    "rolled-back" => true
}}
	at org.jboss.as.test.integration.domain.management.util.DomainTestUtils.executeForResult(DomainTestUtils.java:160)
	at org.jboss.as.test.integration.domain.mixed.DomainAdjuster.applyVersionAdjustments(DomainAdjuster.java:248)
	at org.jboss.as.test.integration.domain.mixed.DomainAdjuster.adjust(DomainAdjuster.java:125)
	at org.jboss.as.test.integration.domain.mixed.DomainAdjuster.adjustForVersion(DomainAdjuster.java:98)
	at org.jboss.as.test.integration.domain.mixed.MixedDomainTestSupport.start(MixedDomainTestSupport.java:79)
	at org.jboss.as.test.integration.domain.mixed.MixedDomainTestSuite.getSupport(MixedDomainTestSuite.java:65)
	at org.jboss.as.test.integration.domain.mixed.eap620.MixedDomain620TestSuite.initializeDomain(MixedDomain620TestSuite.java:43)
------- Stdout: -------
INFO  [org.xnio] XNIO version 3.3.1.Final
INFO  [org.xnio.nio] XNIO NIO Implementation Version 3.3.1.Final
INFO  [org.jboss.remoting] JBoss Remoting version 4.0.9.Final
&amp#27;[0m19:53:37,731 INFO  [org.jboss.modules] (main) JBoss Modules version 1.4.3.Final
&amp#27;[0m&amp#27;[0m19:53:37,933 INFO  [org.jboss.as.process.Host Controller.status] (main) WFLYPC0018: Starting process 'Host Controller'
&amp#27;[0m[Host Controller] &amp#27;[0m19:53:38,440 INFO  [org.jboss.modules] (main) JBoss Modules version 1.4.3.Final&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:38,619 INFO  [org.jboss.msc] (main) JBoss MSC version 1.2.6.Final&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:38,667 INFO  [org.jboss.as] (MSC service thread 1-2) WFLYSRV0049: WildFly Full 10.0.0.Beta2-SNAPSHOT (WildFly Core 2.0.0.Beta2) starting&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:39,186 INFO  [org.xnio] (MSC service thread 1-1) XNIO version 3.3.1.Final&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:39,200 INFO  [org.jboss.as.host.controller] (Controller Boot Thread) WFLYHC0003: Creating http management service using network interface (management) port (9990) securePort (-1)&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:39,209 INFO  [org.xnio.nio] (MSC service thread 1-1) XNIO NIO Implementation Version 3.3.1.Final&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:39,276 INFO  [org.jboss.remoting] (MSC service thread 1-1) JBoss Remoting version 4.0.9.Final&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:39,355 INFO  [org.jboss.as.remoting] (MSC service thread 1-2) WFLYRMT0001: Listening on [::1]:9999&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,392 INFO  [org.jboss.as.controller.management-deprecated] (Controller Boot Thread) WFLYCTL0028: Attribute 'enabled' in the resource at address '/profile=default/subsystem=datasources/data-source=ExampleDS' is deprecated, and may be removed in future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,442 INFO  [org.jboss.as.controller.management-deprecated] (Controller Boot Thread) WFLYCTL0028: Attribute 'enabled' in the resource at address '/profile=ha/subsystem=datasources/data-source=ExampleDS' is deprecated, and may be removed in future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,520 INFO  [org.jboss.as.controller.management-deprecated] (Controller Boot Thread) WFLYCTL0028: Attribute 'enabled' in the resource at address '/profile=full/subsystem=datasources/data-source=ExampleDS' is deprecated, and may be removed in future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,561 INFO  [org.jboss.as.controller.management-deprecated] (Controller Boot Thread) WFLYCTL0028: Attribute 'enabled' in the resource at address '/profile=full-ha/subsystem=datasources/data-source=ExampleDS' is deprecated, and may be removed in future version. See the attribute description in the output of the read-resource-description operation to learn more about the deprecation.&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,897 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0060: Http management interface listening on http://[::1]:9990/management&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,898 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0054: Admin console is not enabled&amp#27;[0m
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:41,898 INFO  [org.jboss.as] (Controller Boot Thread) WFLYSRV0025: WildFly Full 10.0.0.Beta2-SNAPSHOT (WildFly Core 2.0.0.Beta2) (Host Controller) started in 3814ms - Started 43 of 46 services (13 services are lazy, passive or on-demand)&amp#27;[0m
INFO  [org.jboss.as.test.integration.domain.management.util.DomainLifecycleUtil] HostController started in 4844 ms
[Host Controller] &amp#27;[0m&amp#27;[0m19:53:42,174 INFO  [org.jboss.as.controller.management-operation] (management-handler-thread - 4) Adding remove step for child at [&amp#27;[0m
[Host Controller]     ("server-group" => "main-server-group"),
[Host Controller]     ("jvm" => "default")
[Host Controller] ]


org.jboss.as.test.integration.domain.mixed.eap630.MixedDomain630TestSuite: java.lang.AssertionError: expected:<EAP_6_2_0> but was:<EAP_6_3_0>
	at org.jboss.as.test.integration.domain.mixed.MixedDomainTestSuite.getVersion(MixedDomainTestSuite.java:43)
	at org.jboss.as.test.integration.domain.mixed.MixedDomainTestSuite.getSupport(MixedDomainTestSuite.java:55)
	at org.jboss.as.test.integration.domain.mixed.eap630.MixedDomain630TestSuite.initializeDomain(MixedDomain630TestSuite.java:44)


```

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemXMLReader.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemXMLReader.java
@@ -49,7 +49,6 @@ public class JGroupsSubsystemXMLReader implements XMLElementReader<List<ModelNod
         this.schema = schema;
     }
 
-    @SuppressWarnings("deprecation")
     @Override
     public void readElement(XMLExtendedStreamReader reader, List<ModelNode> result) throws XMLStreamException {
 
@@ -218,7 +217,6 @@ public class JGroupsSubsystemXMLReader implements XMLElementReader<List<ModelNod
         }
     }
 
-    @SuppressWarnings("deprecation")
     private void parseStacks(XMLExtendedStreamReader reader, PathAddress address, Map<PathAddress, ModelNode> operations) throws XMLStreamException {
 
         ModelNode operation = operations.get(address);
@@ -339,8 +337,42 @@ public class JGroupsSubsystemXMLReader implements XMLElementReader<List<ModelNod
             }
         }
 
+        for (ThreadPoolResourceDefinition pool : ThreadPoolResourceDefinition.values()) {
+            PathAddress poolAddress = address.append(pool.getPathElement());
+            operations.put(poolAddress, Util.createAddOperation(poolAddress));
+        }
+
         while (reader.hasNext() && (reader.nextTag() != XMLStreamConstants.END_ELEMENT)) {
-            this.parseProtocolElement(reader, address, operations);
+            XMLElement element = XMLElement.forName(reader.getLocalName());
+            switch (element) {
+                case DEFAULT_THREAD_POOL: {
+                    if (this.schema.since(JGroupsSchema.VERSION_3_0)) {
+                        this.parseThreadPool(ThreadPoolResourceDefinition.DEFAULT, reader, address, operations);
+                        break;
+                    }
+                }
+                case INTERNAL_THREAD_POOL: {
+                    if (this.schema.since(JGroupsSchema.VERSION_3_0)) {
+                        this.parseThreadPool(ThreadPoolResourceDefinition.INTERNAL, reader, address, operations);
+                        break;
+                    }
+                }
+                case OOB_THREAD_POOL: {
+                    if (this.schema.since(JGroupsSchema.VERSION_3_0)) {
+                        this.parseThreadPool(ThreadPoolResourceDefinition.OOB, reader, address, operations);
+                        break;
+                    }
+                }
+                case TIMER_THREAD_POOL: {
+                    if (this.schema.since(JGroupsSchema.VERSION_3_0)) {
+                        this.parseThreadPool(ThreadPoolResourceDefinition.TIMER, reader, address, operations);
+                        break;
+                    }
+                }
+                default: {
+                    this.parseProtocolElement(reader, address, operations);
+                }
+            }
         }
     }
 
@@ -390,18 +422,6 @@ public class JGroupsSubsystemXMLReader implements XMLElementReader<List<ModelNod
                 this.parseProperty(reader, address, operations);
                 break;
             }
-            case DEFAULT_THREAD_POOL:
-                parseThreadPool(ThreadPoolResourceDefinition.DEFAULT, reader, address, operations);
-                break;
-            case INTERNAL_THREAD_POOL:
-                parseThreadPool(ThreadPoolResourceDefinition.INTERNAL, reader, address, operations);
-                break;
-            case OOB_THREAD_POOL:
-                parseThreadPool(ThreadPoolResourceDefinition.OOB, reader, address, operations);
-                break;
-            case TIMER_THREAD_POOL:
-                parseThreadPool(ThreadPoolResourceDefinition.TIMER, reader, address, operations);
-                break;
             default: {
                 throw ParseUtils.unexpectedElement(reader);
             }
@@ -416,8 +436,7 @@ public class JGroupsSubsystemXMLReader implements XMLElementReader<List<ModelNod
 
     private void parseThreadPool(ThreadPoolResourceDefinition pool, XMLExtendedStreamReader reader, PathAddress parentAddress, Map<PathAddress, ModelNode> operations) throws XMLStreamException {
         PathAddress address = parentAddress.append(pool.getPathElement());
-        ModelNode operation = Util.createAddOperation(address);
-        operations.put(address, operation);
+        ModelNode operation = operations.get(address);
 
         for (int i = 0; i < reader.getAttributeCount(); i++) {
             XMLAttribute attribute = XMLAttribute.forName(reader.getAttributeLocalName(i));

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemXMLWriter.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/JGroupsSubsystemXMLWriter.java
@@ -22,11 +22,13 @@
 
 package org.jboss.as.clustering.jgroups.subsystem;
 
+import java.util.Collection;
 import java.util.EnumSet;
 
 import javax.xml.stream.XMLStreamException;
 
 import org.jboss.as.clustering.controller.Attribute;
+import org.jboss.as.controller.PathElement;
 import org.jboss.as.controller.persistence.SubsystemMarshallingContext;
 import org.jboss.dmr.ModelNode;
 import org.jboss.dmr.Property;
@@ -132,11 +134,14 @@ public class JGroupsSubsystemXMLWriter implements XMLElementWriter<SubsystemMars
     }
 
     private static void writeThreadPoolElements(XMLElement element, ThreadPoolResourceDefinition pool, XMLExtendedStreamWriter writer, ModelNode transport) throws XMLStreamException {
-        if (transport.get(pool.getPathElement().getKey()).hasDefined(pool.getPathElement().getValue())) {
-            ModelNode threadPool = transport.get(pool.getPathElement().getKeyValuePair());
-            writer.writeStartElement(element.getLocalName());
-            writeAttributes(writer, threadPool, pool.getAttributes());
-            writer.writeEndElement();
+        PathElement path = pool.getPathElement();
+        if (transport.get(path.getKey()).hasDefined(path.getValue())) {
+            ModelNode threadPool = transport.get(path.getKeyValuePair());
+            if (hasDefined(threadPool, pool.getAttributes())) {
+                writer.writeStartElement(element.getLocalName());
+                writeAttributes(writer, threadPool, pool.getAttributes());
+                writer.writeEndElement();
+            }
         }
     }
 
@@ -154,11 +159,15 @@ public class JGroupsSubsystemXMLWriter implements XMLElementWriter<SubsystemMars
         writer.writeEndElement();
     }
 
+    private static boolean hasDefined(ModelNode model, Collection<? extends Attribute> attributes) {
+        return attributes.stream().anyMatch(attribute -> model.hasDefined(attribute.getDefinition().getName()));
+    }
+
     private static <A extends Enum<A> & Attribute> void writeAttributes(XMLExtendedStreamWriter writer, ModelNode model, Class<A> attributeClass) throws XMLStreamException {
         writeAttributes(writer, model, EnumSet.allOf(attributeClass));
     }
 
-    private static void writeAttributes(XMLExtendedStreamWriter writer, ModelNode model, Iterable<? extends Attribute> attributes) throws XMLStreamException {
+    private static void writeAttributes(XMLExtendedStreamWriter writer, ModelNode model, Collection<? extends Attribute> attributes) throws XMLStreamException {
         for (Attribute attribute : attributes) {
             writeAttribute(writer, model, attribute);
         }

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/ThreadPoolResourceDefinition.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/subsystem/ThreadPoolResourceDefinition.java
@@ -23,6 +23,7 @@
 package org.jboss.as.clustering.jgroups.subsystem;
 
 import java.util.Arrays;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 
@@ -148,7 +149,7 @@ public enum ThreadPoolResourceDefinition implements ResourceDefinition, Registra
         registration.registerSubModel(this);
     }
 
-    Iterable<Attribute> getAttributes() {
+    Collection<Attribute> getAttributes() {
         return Arrays.asList(this.minThreads, this.maxThreads, this.queueLength, this.keepAliveTime);
     }
 

--- a/clustering/jgroups/extension/src/test/java/org/jboss/as/clustering/jgroups/subsystem/SubsystemParsingTestCase.java
+++ b/clustering/jgroups/extension/src/test/java/org/jboss/as/clustering/jgroups/subsystem/SubsystemParsingTestCase.java
@@ -70,10 +70,10 @@ public class SubsystemParsingTestCase extends ClusteringSubsystemTest {
     @Parameters
     public static Collection<Object[]> data() {
         Object[][] data = new Object[][] {
-                { JGroupsSchema.VERSION_1_1, 20 },
-                { JGroupsSchema.VERSION_2_0, 22 },
-                { JGroupsSchema.VERSION_3_0, 29 },
-                { JGroupsSchema.VERSION_4_0, 29 },
+                { JGroupsSchema.VERSION_1_1, 28 },
+                { JGroupsSchema.VERSION_2_0, 30 },
+                { JGroupsSchema.VERSION_3_0, 33 },
+                { JGroupsSchema.VERSION_4_0, 33 },
         };
         return Arrays.asList(data);
     }

